rand_distr/src/hypergeometric.rs

```diff
@@ -0,0 +1,404 @@
+            Error::PopulationTooLarge => "total_population_size is too large causing underflow in geometric distribution",
+            Error::ProbabilityTooLarge => "population_with_feature > total_population_size in geometric distribution",
+            Error::SampleSizeTooLarge => "sample_size > total_population_size in geometric distribution",
+        })
+    }
+}
+
+// evaluate fact(numerator.0)*fact(numerator.1) / fact(denominator.0)*fact(denominator.1)
+fn fraction_of_products_of_factorials(numerator: (u64, u64), denominator: (u64, u64)) -> f64 {
+    let min_top = u64::min(numerator.0, numerator.1);
+    let min_bottom = u64::min(denominator.0, denominator.1);
+    // the factorial of this will cancel out:
+    let min_all = u64::min(min_top, min_bottom);
+
+    let max_top = u64::max(numerator.0, numerator.1);
+    let max_bottom = u64::max(denominator.0, denominator.1);
+    let max_all = u64::max(max_top, max_bottom);
+
+    let mut result = 1.0;
+    for i in (min_all + 1)..=max_all {
+        if i <= min_top {
+            result *= i as f64;
+        }
+        
+        if i <= min_bottom {
+            result /= i as f64;
+        }
+        
+        if i <= max_top {
+            result *= i as f64;
+        }
+        
+        if i <= max_bottom {
+            result /= i as f64;
+        }
+    }
+    
+    result
+}
+
+fn ln_of_factorial(v: f64) -> f64 {
+    // the paper calls for ln(v!), but also wants to pass in fractions,
+    // so we need to use Stirling's approximation to fill in the gaps:
+    v * v.ln() - v
+}
+
+impl Hypergeometric {
+    /// Constructs a new `Hypergeometric` with the given shape parameters.
+    #[allow(clippy::many_single_char_names)] // Same names as in the reference.
+    pub fn new(total_population_size: u64, population_with_feature: u64, sample_size: u64) -> Result<Self, Error> {
+        if population_with_feature > total_population_size {
+            return Err(Error::ProbabilityTooLarge);
+        }
+
+        if sample_size > total_population_size {
+            return Err(Error::SampleSizeTooLarge);
+        }
+
+        // set-up constants as function of original parameters
+        let n = total_population_size;
+        let (mut sign_x, mut offset_x) = (1, 0);
+        let (n1, n2) = {
+            // switch around success and failure states if necessary to ensure n1 <= n2
+            let population_without_feature = n - population_with_feature;
+            if population_with_feature > population_without_feature {
+                sign_x = -1;
+                offset_x = sample_size as i64;
+                (population_without_feature, population_with_feature)
+            } else {
+                (population_with_feature, population_without_feature)
+            }
+        };
+        // when sampling more than half the total population, take the smaller
+        // group as sampled instead (we can then return n1-x instead):
+        let k = if sample_size <= n / 2 {
```

Comment:
I think this should be `<` according to the paper.

Reply:
Maybe so, I'm confident this is right, though.
When `n` is even, it doesn't make a difference whether we switch when `k == n/2`.
When `n` is odd, `n/2 < n - n/2`, so if we switch when `k == n/2`, we're actually taking the bigger group as sampled, which runs counter to the stated intent.